### PR TITLE
Gas optimization

### PR DESCRIPTION
**Pull request by mrsmkl**
## Description

What could be removed
* some parameters from conditions can be removed, this will make creating agreements at least 50% cheaper
* in theory creating an agreement would need only one or perhaps no on-chain variables, but fulfilling the conditions still requires gas so there's probably no point in optimizing it so much
* access agreement when condition creation is optimized: 610k (originally 1040k)
* creating access agreement when agreement creation is optimized: 300k (also removed condition ids from storage)
* lock 120k + approve 40k
* fulfill escrow 100k
* access condition 90k
* snark proof: 300k
* creating condition: 110k (originally 230k)

Questions:
* should we keep condition ids in agreements (otherwise can be checked from events)
* do we need provenance in storage or are events enough?

Adds a DoS protection for agreements similar to DIDs.

## Is this PR related with an open issue?

Related to Issue #236

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Checklist:

- [ ] Follows the code style of this project.
- [ ] Tests Cover Changes
- [ ] Documentation

#### Funny gif

Put a link of a funny gif inside the parenthesis-->

### CODE REVIEW

File: contracts/agreements/AgreementStoreLibrary.sol [hunk 1]
```diff
@@ -20,18 +20,18 @@ pragma solidity ^0.8.0;
 library AgreementStoreLibrary {
 
     struct Agreement {
-        bytes32 did;
+        bytes32 did; // UNUSED
```
Comment (aaitor) [inline contracts/agreements/AgreementStoreLibrary.sol:23]: If we are not using them, shouldn't we remove?
If it's for compatibility purposes we can merge this into a `2.x` or `unstable` branch

File: contracts/conditions/ConditionStoreLibrary.sol [hunk 2]
```diff
@@ -24,15 +24,15 @@ library ConditionStoreLibrary {
     struct Condition {
         address typeRef;
         ConditionState state;
-        address createdBy;
-        address lastUpdatedBy;
-        uint256 blockNumberUpdated;
+        address createdBy; // UNUSED
```
Comment (aaitor) [inline contracts/conditions/ConditionStoreLibrary.sol:27]: Same, if not used let's remove them

File: contracts/conditions/LockPaymentCondition.sol [hunk 3]
```diff
@@ -159,19 +159,23 @@ contract LockPaymentCondition is ILockPayment, ReentrancyGuardUpgradeable, Condi
             _agreementId,
             hashValues(_did, _rewardAddress, _tokenAddress, _amounts, _receivers)
         );
+        
+        // return ConditionStoreLibrary.ConditionState.Fulfilled;
```
Comment (aaitor) [inline contracts/conditions/LockPaymentCondition.sol:163]: cleanup

File: contracts/conditions/LockPaymentCondition.sol [hunk 4]
```diff
@@ -240,14 +244,15 @@ contract LockPaymentCondition is ILockPayment, ReentrancyGuardUpgradeable, Condi
             ConditionStoreLibrary.ConditionState.Fulfilled
         );
 
+/*
```
Comment (aaitor) [inline contracts/conditions/LockPaymentCondition.sol:247]: cleanup

File: contracts/libraries/EpochLibrary.sol [hunk 5]
```diff
@@ -31,7 +31,7 @@ library EpochLibrary {
 
     struct EpochList {
         mapping(bytes32 => Epoch) epochs;
-        bytes32[] epochIds;
+        bytes32[] epochIds; // UNUSED
```
Comment (aaitor) [inline contracts/libraries/EpochLibrary.sol:34]: remove if not used

File: contracts/registry/DIDFactory.sol [hunk 6]
```diff
@@ -652,25 +656,25 @@ contract DIDFactory is OwnableUpgradeable, ProvenanceRegistry {
...
      * @return size the length of the DID registry.
-     */
     function getDIDRegistrySize()
     public
     view
     returns (uint size)
     {
         return didRegisterList.didRegisterIds.length;
     }
+     */
```
Comment (aaitor) [inline contracts/registry/DIDFactory.sol:666]: cleanup

File: contracts/registry/DIDRegistryLibrary.sol [hunk 7]
```diff
@@ -48,7 +48,7 @@ library DIDRegistryLibrary {
     // List of DID's registered in the system
     struct DIDRegisterList {
         mapping(bytes32 => DIDRegister) didRegisters;
-        bytes32[] didRegisterIds;
+        bytes32[] didRegisterIds; // UNUSED
```
Comment (aaitor) [inline contracts/registry/DIDRegistryLibrary.sol:51]: remove if not used

File: contracts/templates/BaseEscrowTemplate.sol [hunk 8]
```diff
@@ -100,51 +99,69 @@ contract BaseEscrowTemplate is AgreementTemplate {
...
         if (providers.length > 0) {
             agreementData.agreementDataItems[_id]
                 .accessProvider = providers[0];
         } else {
             agreementData.agreementDataItems[_id]
                 .accessProvider = owner;
-        }
-
-        agreementData.agreementIds.push(_id);
+        }*/
```
Comment (aaitor) [inline contracts/templates/BaseEscrowTemplate.sol:152]: cleanup

File: contracts/templates/BaseEscrowTemplate.sol [hunk 9]
```diff
@@ -163,7 +180,22 @@ contract BaseEscrowTemplate is AgreementTemplate {
...
+        (owner, , , , , providers,,,) = didRegistryInstance.getDIDRegister(agreementData.agreementDataItems[_id].did);
+
+        if (providers.length > 0) {
+            accessProvider = providers[0];
+        } else {
+            accessProvider = owner;
+        }
         accessConsumer = agreementData.agreementDataItems[_id].accessConsumer;
-        accessProvider = agreementData.agreementDataItems[_id].accessProvider;
+        // accessProvider = agreementData.agreementDataItems[_id].accessProvider;
```
Comment (aaitor) [inline contracts/templates/BaseEscrowTemplate.sol:199]: cleanup

File: contracts/test/DIDRegistry/DIDRegistryChangeFunctionSignature.sol [hunk 10]
```diff
@@ -48,6 +47,6 @@ contract DIDRegistryChangeFunctionSignature is DIDRegistry {
             block.number
         );
         
-        return getDIDRegistrySize();
+        // return getDIDRegistrySize();
```
Comment (aaitor) [inline contracts/test/DIDRegistry/DIDRegistryChangeFunctionSignature.sol:50]: cleanup

File: test/int/agreement/DynamicAccessAgreement.Test.js [hunk 11]
```diff
@@ -165,33 +172,31 @@ contract('Dynamic Access Template integration test', (accounts) => {
...
 
             // create agreement
-            await dynamicAccessTemplate.createAgreement(agreementId, ...Object.values(agreement))
+            await dynamicAccessTemplate.createAgreement(...Object.values(agreement))
 
             // check state of agreement and conditions
-            const _did = await didRegistry.hashDID(constants.did[0], receiver)
-            expect((await agreementStoreManager.getAgreement(agreementId)).did)
-                .to.equal(_did)
+            // const _did = await didRegistry.hashDID(constants.did[0], receiver)
```
Comment (aaitor) [inline test/int/agreement/DynamicAccessAgreement.Test.js:178]: cleanup

File: test/int/nft/NFT721_e2e.Test.js [hunk 12]
```diff
@@ -33,8 +33,9 @@ contract('End to End NFT721 Scenarios', (accounts) => {
     const royalties = 10 // 10% of royalties in the secondary market
     const didSeed = testUtils.generateId()
     let did
-    const agreementId = testUtils.generateId()
-    const agreementId2 = testUtils.generateId()
+    let agreementId
+    // const agreementId = testUtils.generateId()
```
Comment (aaitor) [inline test/int/nft/NFT721_e2e.Test.js:37]: cleanup

File: test/int/nft/NFT721_e2e.Test.js [hunk 13]
```diff
@@ -300,25 +311,26 @@ contract('End to End NFT721 Scenarios', (accounts) => {
             it('The artist can check the payment and transfer the NFT to the collector', async () => {
                 await nft.setApprovalForAll(transferCondition.address, true, { from: artist })
 
+                /*
```
Comment (aaitor) [inline test/int/nft/NFT721_e2e.Test.js:314]: cleanup

**Review by aaitor — COMMENTED** [review]
It's looking good. There is pending a comments clean up and removing unused stuff.
That would lead to break compatibility.
I suggest apply these changes and merge into a `2.x` contracts branch. Apply also the version `2.0.0-rc0` or `1.99.0`

File: contracts/agreements/AgreementStoreLibrary.sol (hunk 1 above)
Comment (mrsmkl) [reply contracts/agreements/AgreementStoreLibrary.sol:23]: I'm keeping them for contract upgradability, it will be lost if they are removed...

**Review by aaitor — APPROVED** [review]
LGTM!

**Pull request merged**